pinot-core/src/main/java/org/apache/pinot/core/data/manager/realtime/MaximumPinotConsumptionDelayTracker.java

```diff
@@ -0,0 +1,344 @@
+    if ((maxNumberOfTicksBeforeVerification < 0) || (initialTimeThreadDelayMs < 0)
+        || (timerThreadTickIntervalMs <= 0)) {
+      throw new RuntimeException("Illegal timer arguments");
+    }
+    _enableAging = true;
+    _maxNumberOfTicksBeforeVerification = maxNumberOfTicksBeforeVerification;
+    _initialTimeThreadDelayMs = initialTimeThreadDelayMs;
+    _timerThreadTickIntervalMs = timerThreadTickIntervalMs;
+    _timerTask = new TrackingTimerTask("TimerTask" + tableNameWithType, this);
+    _timer = new Timer();
+    _timer.schedule(_timerTask, _initialTimeThreadDelayMs, _timerThreadTickIntervalMs);
+    // Install callback metric
+    _serverMetrics.addCallbackTableGaugeIfNeeded(_tableNameWithType, ServerGauge.MAX_PINOT_CONSUMPTION_DELAY_MS,
+        () -> (long) getMaxPinotConsumptionDelay());
+  }
+
+  // Constructor that uses default timeout
+  public MaximumPinotConsumptionDelayTracker(ServerMetrics sm, String tableNameWithType,
+      RealtimeTableDataManager tdm) {
+    this(sm, tableNameWithType, tdm, MAX_NUMBER_OF_TIME_TICKS_BEFORE_VERIFICATION,
+        INITIAL_TIMER_THREAD_DELAY_MS, TIMER_THREAD_TICK_INTERVAL_MS);
+  }
+
+  /**
+   * Use to set or rest the aging of reported values.
+   * @param enableAging true if we want maximum to be aged as per sample time or false if we do not want to age
+   *                   samples
+   */
+  public void setEnableAging(boolean enableAging) {
+    _enableAging = enableAging;
+  }
+
+  /*
+   * Called by LLRealTimeSegmentDataManagers to post delay updates to this tracker class.
+   * If the new sample represents a new Maximum we update the current maximum.
+   * If the new sample was for the partition that was maximum, but delay is not maximum anymore, we must select
+   * a new maximum.
+   * Make it synchronize method so that it can be called concurrently from multiple LLRealTimeSegmentDataManagers.
+   *
+   * @param delayInMilliseconds pinot consumption delay being recorded.
+   * @param currentTime sample time.
+   * @param partitionId partition ID for which this delay is being recorded.
+   */
+  synchronized public void recordPinotConsumptionDelay(long delayInMilliseconds, long currentTime, long partitionId) {
+    if (delayInMilliseconds > _maxPinotConsumptionDelay) {
+      _maxPinotConsumptionDelay = delayInMilliseconds;
+      _maxPartitionId = partitionId;
+    }
+    // Store new measure and wipe old one for this partition
+    _partitionToDelaySampleMap.put(partitionId, new DelayMeasure(currentTime, delayInMilliseconds));
+    if (_partitionsMarkedForConfirmation.containsKey(partitionId)) {
+      // If we are consuming we do not need to track this partition for removal.
+      _partitionsMarkedForConfirmation.remove(partitionId);
+    }
+    if ((partitionId == _maxPartitionId) && (delayInMilliseconds < _maxPinotConsumptionDelay)) {
+      // If the partition that was maximum stops being maximum, we need to select a new maximum.
+      updateMaximum();
+    }
```

Comment:
The frequency of getting the gauge metric value might be in minutes while the frequency of updating `_partitionToDelaySampleMap` is in sub-second. I suggest not to track maxDelay and maxPartitionId when this method is called for each partition consumer. We can simply just update `_partitionToDelaySampleMap` in this method. When the call comes to fetch the gauge metric value, we can loop over the `_partitionToDelaySampleMap` and find the maximum.

Reply:
done